hasher.go

```diff
@@ -159,21 +164,42 @@ func CalculateLimit(maxCapacity, numItems, size uint64) uint64 {
 
 func (h *Hasher) FillUpTo32() {
 	// pad zero bytes to the left
-	if rest := len(h.buf) % 32; rest != 0 {
-		h.buf = append(h.buf, zeroBytes[:32-rest]...)
+	if h.pos != 0 {
+		copy(h.buf[len(h.buf)-1][h.pos:], zeroBytes[:32-h.pos])
 	}
+	h.pos = 0
 }
 
 func (h *Hasher) AppendUint8(i uint8) {
-	h.buf = MarshalUint8(h.buf, i)
+	dst := make([]byte, 0)
+	dst = MarshalUint8(dst, i)
+	if h.pos == 0 {
+		var b32 [32]byte
+		copy(b32[:], dst)
+		h.buf = append(h.buf, b32)
+	} else {
+		copy(h.buf[len(h.buf)-1][h.pos:], dst)
+	}
+	h.pos = (h.pos + 1) % 32
 }
 
 func (h *Hasher) AppendUint64(i uint64) {
-	h.buf = MarshalUint64(h.buf, i)
+	dst := make([]byte, 0)
```

Comment:
Since these functions are performance critical I would go with something like 
```go
func (h *Hasher) AppendUint64(i uint64) {
    if h.pos = 0 {
        h.buf = append(h.buf, [32]byte{})
    }
    j := 0
    for; j < 8 && j < 32 - h.pos & ; j++ {
        h.buf[len(h.buf) - 1][h.pos + j] = byte(i >> 8 * j)
    }
    // Deal with overflows
    h.pos = (h.pos + j) % 32
    if h.pos == 0 && j < 8 {
        h.buf = append(h.buf, [32]byte{})
        for; j < 8 & ; j++ {
            h.buf[len(h.buf) - 1][h.pos] = byte(i >> 8 * j)
            h.pos++
        }
    }
}
```